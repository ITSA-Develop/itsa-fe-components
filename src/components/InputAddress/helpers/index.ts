// @ts-nocheck
export const getPostalCodeFromLatLng = (lat: number, lng: number): void => {
	const geocoder = new google.maps.Geocoder();
	const latlng: google.maps.LatLngLiteral = { lat, lng };

	geocoder.geocode(
		{ location: latlng },
		(results: google.maps.GeocoderResult[] | null, status: google.maps.GeocoderStatus) => {
			if (status === 'OK' && results && results.length > 0) {
				const postalCodeComponent = results[0].address_components.find(comp => comp.types.includes('postal_code'));

				if (postalCodeComponent) {
					console.log('Código postal:', postalCodeComponent.long_name);
				} else {
					console.log('No se encontró código postal para esta ubicación.');
				}
			} else {
				console.error('Geocoder falló por:', status);
			}
		},
	);
};
