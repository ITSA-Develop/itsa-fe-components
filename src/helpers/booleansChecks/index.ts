import { REGEX_FORMATS } from '@/constants';

export const isArray = <T = unknown>(value: unknown): value is T[] => Array.isArray(value);

export const isValidEmailFormat = (email?: string): boolean => !!email && REGEX_FORMATS.email.test(email);

export const isEmpty = (value: unknown): boolean => {
	if (value == null) return true;
	if (typeof value === 'string') return value.trim().length === 0;
	if (Array.isArray(value)) return value.length === 0;
	if (value instanceof Map || value instanceof Set) return value.size === 0;
	if (isPlainObject(value)) return Object.keys(value).length === 0;
	return false;
};

export const isValidMaskFormat = (inputValue?: string): boolean => !!inputValue && !inputValue.includes('_');

export const isNumber = (value: unknown): value is number => typeof value === 'number' && Number.isFinite(value);

export const isObject = (value: unknown): value is object => typeof value === 'object' && value !== null;

export const isPlainObject = (value: unknown): value is Record<string, unknown> =>
	Object.prototype.toString.call(value) === '[object Object]';

export const isString = (value: unknown): value is string => typeof value === 'string';

export const isUndefined = (value: unknown): value is undefined => value === undefined;
