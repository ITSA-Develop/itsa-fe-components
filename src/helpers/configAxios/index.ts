// En un archivo como: src/api/securityApi.ts
import axios, { type AxiosInstance, type InternalAxiosRequestConfig } from 'axios';
// import { useAuthStore } from '@/store/auth.store'; // 1. Importa tu store
import { EMicroFrontends } from '@/enums';
import { clearLocalStorage } from '@/index';


// Puedes crear la instancia base aquí
const initialInstance = axios.create();

// La función que configura la instancia
export const getInstance = (instance: AxiosInstance) => {
	instance.interceptors.request.use(
		async (config: InternalAxiosRequestConfig) => {
			// La configuración se mueve aquí adentro
			// const apiVersion = 'v1'; // O la versión que necesites
			const baseURL = import.meta.env.VITE_API;	
			const finalBaseUrl = `${baseURL}`;

			// 2. Obtén el token del store en cada petición
			const token = null; // useAuthStore.getState().token;

			// Asigna la nueva configuración
			config.baseURL = finalBaseUrl;
			if (config.headers && token) {
				config.headers.Authorization = `Bearer ${token}`;
			}

			return config;
		},
		error => {
			// Es bueno manejar errores aquí también
			return Promise.reject(error);
		},
	);

	instance.interceptors.response.use(
		response => response,
		async error => {
			// Aquí puedes manejar errores globales, como un 401 (token expirado)
			// y ejecutar la lógica de logout o de refrescar el token.
			return Promise.reject(error);
		},
	);

	return instance;
};

// Exporta la instancia ya configurada
export const securityApi = getInstance(initialInstance);

export const logoutAndClearLocalStorage = () => {
	clearLocalStorage();
	window.location.href = EMicroFrontends.itsaBackOffice + '/login';
};
