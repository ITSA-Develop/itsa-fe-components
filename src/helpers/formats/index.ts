import { TInputOptions } from '@/types';
import { IEntitySchema, TFieldDisplayType } from '@/types/schema';

export const getDropdownFormattedOptions = (optionValues: Record<any, string>) => {
	return Object.entries(optionValues).map(([key, value]) => ({
		value: key,
		label: value,
	}));
};

export const getFormattedAddress = ({
	principalStreet,
	streetNumber,
	parish,
	canton,
	province,
}: {
	principalStreet: string;
	streetNumber: string;
	parish?: string;
	canton: string;
	province: string;
}): string => {
	const parts = [`${principalStreet} ${streetNumber}`, parish, canton, province];
	return parts.filter(Boolean).join(', ');
};

const formatters: Partial<Record<TFieldDisplayType, (value: any, populateValue?: any) => string>> = {
	text: value => value,
	dropdown: (value, populateValue) =>
		populateValue ? (populateValue.find((populate: any) => populate.value === value)?.label ?? '') : '',
	radio: value => value?.label ?? '',
	currency: value => value,
	checkbox: value => (value ? 'SÃ­' : 'No'),
};

export const getFormattedFieldValue = (key: TFieldDisplayType, value: any, populateValue: any): string => {
	const formatter = formatters[key];
	return formatter ? formatter(value, populateValue) : String(value);
};

export const getFormattedSchemaValues = (
	elem: Record<string, any>,
	schema: IEntitySchema,
	populate?: Record<string, TInputOptions[]>,
): Record<string, string> => {
	return Object.entries(elem).reduce(
		(acc, [field, value]) => {
			const fieldSchema = schema.fields[field as keyof typeof schema.fields];

			if (fieldSchema) {
				const options = populate?.[field as keyof typeof populate] ?? null;

				acc[field] = getFormattedFieldValue(fieldSchema.type as TFieldDisplayType, value, options);
			}

			return acc;
		},
		{} as Record<string, string>,
	);
};

export const getSchemaColumns = (schema: IEntitySchema): string[] => {
	return Object.values(schema.fields)
		.filter(config => !config.excludeFromColumns)
		.map(config => config.label)
		.filter((label): label is string => typeof label === 'string')
		.concat('actions');
};

export const getSchemaFields = (schema: IEntitySchema): string[] => {
	return Object.entries(schema.fields)
		.filter(([, config]) => !config.excludeFromFields)
		.map(([key]) => key)
		.concat('actions');
};

export const getLabelFromValue = (value: number | string, options: TInputOptions[]): string => {
	const match = options?.find(opt => opt?.value === value);
	return match?.label ?? '';
};


export function getNumberFromStorage(key: string, defaultValue = 0): number {
	try {
		const raw = localStorage.getItem(key);
		if (!raw) return defaultValue;

		const num = Number(raw);
		return Number.isNaN(num) ? defaultValue : num;
	} catch {
		return defaultValue;
	}
}

export function getStringFromStorage(key: string, defaultValue = ''): string {
	try {
		const raw = localStorage.getItem(key);
		return raw ?? defaultValue;
	} catch {
		return defaultValue;
	}
}

