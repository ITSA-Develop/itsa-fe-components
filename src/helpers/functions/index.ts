import { TNotificationProps } from '@/types';
import { notification } from 'antd';

export const openNotificationWithIcon = ({ type, message, description }: TNotificationProps) => {
	notification[type]({
		message,
		description,
	});
};
