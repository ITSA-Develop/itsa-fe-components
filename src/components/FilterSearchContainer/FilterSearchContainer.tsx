import { Button } from '@/components/Button/Button';
import { EButtonColor } from '@/enums';
import { clearURLParams } from '@/helpers';
import { ReactNode } from 'react';

export interface IFilterSearchContainerProps {
	handleSubmit: (data: unknown) => void; // TODO: define when new UI implementation starts
	reset: () => void; // TODO: define when new UI implementation starts
	children: ReactNode;
}

export const FilterSearchContainer = ({ handleSubmit, reset, children }: IFilterSearchContainerProps) => {
	const handleClear = () => {
		reset();
		clearURLParams();
	};

	return (
		<form onSubmit={handleSubmit}>
			<div className="p-4 w-full bg-grey-100 rounded-md">
				<div className="flex flex-wrap gap-4 items-center justify-between">
					{children}
					<Button onClick={handleClear} color={EButtonColor.primary}>
						Limpiar Filtros
					</Button>
					{/* To trigger the onsubmit */}
					<Button htmlType="submit" className="!hidden" />
				</div>
			</div>
		</form>
	);
};
