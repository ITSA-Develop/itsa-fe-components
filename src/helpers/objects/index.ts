import { ItemType, MenuItemType } from 'antd/es/menu/interface';
import { IMenuItem, IPermissionSubmodule } from '../../interfaces';
import { TExtendedMenuItem, TInputOptions } from '../../types';
import { ICON_MAP } from '../../utils/constants';
import { isEmpty } from '../booleansChecks';
import { toTitleCase } from '../strings';
import { ELocalStorageKeys } from '@/enums';

type Tree = Record<string, any>;

const EMPTY_DEFAULT_INPUT_LABEL = { value: '', label: '' };

export const cleanObject = (obj: any) => {
	if (typeof obj !== 'object' || obj === null) return obj;

	const result: any = Array.isArray(obj) ? [] : {};

	for (const key in obj) {
		const value = obj[key];

		if (typeof value === 'object' && value !== null) {
			const nested = cleanObject(value);
			if (!isEmpty(nested)) {
				result[key] = nested;
			}
		} else if (!isEmpty(value)) {
			result[key] = value;
		}
	}

	return result;
};

export const getMappedOptionFromLocal = (
	value: string | null | undefined,
	options: Record<string, string>,
): TInputOptions => {
	return value ? { label: options[value] ?? value, value } : { value: '', label: '' };
};

export const createValueLabelMap = (options: TInputOptions[]) =>
	options.reduce(
		(acc, curr) => {
			acc[curr.value] = curr.label;
			return acc;
		},
		{} as Record<string, string>,
	);

export const getMappedOptionFromAPI = (
	value: string | undefined | null | number,
	options: TInputOptions[],
): TInputOptions => {
	if (!value) return EMPTY_DEFAULT_INPUT_LABEL;

	const map = createValueLabelMap(options);
	const label = map[value];

	return label ? { value, label } : EMPTY_DEFAULT_INPUT_LABEL;
};

export const mapPermissionsToMenuFormat = (subModules: IPermissionSubmodule[]): IMenuItem[] => {
	return subModules.map((module): IMenuItem => {
		const subList: IMenuItem[] = [];

		if (module.groups?.length) {
			const groupItems = module.groups.map(
				(group): IMenuItem => ({
					id: group.id,
					name: group.name,
					icon: ICON_MAP[group.icon as keyof typeof ICON_MAP],
					path: group.path,
					title: toTitleCase(group.name),
					subList: group?.programs
						? group.programs.map(
								(program): IMenuItem => ({
									id: program.id,
									name: program.name,
									title: toTitleCase(program.name),
									icon: ICON_MAP[program.icon as keyof typeof ICON_MAP],
									path: program.path,
									actions: program.actions,
								}),
							)
						: [],
				}),
			);

			subList.push(...groupItems);
		}

		if (module.programs?.length) {
			const directPrograms = module.programs.map(
				(program): IMenuItem => ({
					id: program.id,
					name: program.name,
					title: toTitleCase(program.name),
					icon: ICON_MAP[program.icon as keyof typeof ICON_MAP],
					path: program.path,
					actions: program.actions,
				}),
			);

			subList.push(...directPrograms);
		}

		return {
			id: module.id,
			name: module.name,
			title: toTitleCase(module.name),
			icon: ICON_MAP[module.icon as keyof typeof ICON_MAP],
			path: module.path,
			subList,
		};
	});
};

export const filterMenuTree = (nodes: IMenuItem[], searchTerm: string): IMenuItem[] => {
	const normalize = (text: string) =>
		text
			.normalize('NFD')
			.replace(/[\u0300-\u036f]/g, '')
			.toLowerCase();

	const normalizedSearch = normalize(searchTerm);

	return nodes
		.map(node => {
			const newNode: IMenuItem = { ...node };

			if (newNode.subList && Array.isArray(newNode.subList)) {
				newNode.subList = filterMenuTree(newNode.subList, searchTerm);
			}

			const matchesTitle = newNode.title ? normalize(newNode.title).includes(normalizedSearch) : false;

			const hasMatchingChildren = newNode.subList && newNode.subList.length > 0;

			if (matchesTitle || hasMatchingChildren) {
				return newNode;
			}

			return null;
		})
		.filter((node): node is IMenuItem => node !== null);
};

export const generateFilteredMenu = (
	subModules: IPermissionSubmodule[] | undefined,
	searchTerm: string,
): IMenuItem[] => {
	if (!subModules) return [];

	const mappedMenu = mapPermissionsToMenuFormat(subModules);

	if (searchTerm.trim() !== '') {
		return filterMenuTree(mappedMenu, searchTerm);
	}

	return mappedMenu;
};

export const hasAllRequiredFields = (values: Record<string, any>, requiredFields: string[]): boolean => {
	return requiredFields.every(field => {
		const value = values[field];

		if (value === null || value === undefined) return false;

		if (typeof value === 'string' && value.trim() === '') return false;

		if (Array.isArray(value) && value.length === 0) return false;

		return true;
	});
};

export const findLastChild = (tree: Tree, nodeId: string): any => {
	return tree?.[nodeId] ? findLastChild(tree[nodeId], nodeId) : tree;
};

export const getNestedValue = <T, U = undefined>(obj: T, path: string | string[], defaultValue?: U): U | any => {
	const keys = Array.isArray(path) ? path : path.split('.');
	let result: any = obj;

	for (const key of keys) {
		if (result && Object.prototype.hasOwnProperty.call(result, key)) {
			result = result[key];
		} else {
			return defaultValue;
		}
	}

	return result;
};

export const filterMenuSidebar = (items: MenuItemType[], query: string): MenuItemType[] => {
	const norm = (s: string) =>
		s
			.toLowerCase()
			.normalize('NFD')
			.replace(/\p{Diacritic}/gu, '');

	const getText = (n: React.ReactNode) => (typeof n === 'string' || typeof n === 'number' ? String(n) : '');

	const q = norm(query.trim());

	if (!q) {
		return items;
	}

	// New object functions from here
	const filterNode = (node: ItemType<MenuItemType>): ItemType<MenuItemType> | null => {
		if (!node) {
			return null;
		}

		const nodeLabel = (node as any).label;
		const nodeLabelUppercase = (nodeLabel ?? '').toUpperCase();
		const selfMatch = norm(getText(nodeLabelUppercase)).includes(q) || nodeLabelUppercase.includes(q.toUpperCase());
		const nodeChildren = (node as any).children;

		if (nodeChildren && Array.isArray(nodeChildren) && nodeChildren.length > 0) {
			const filteredChildren = nodeChildren
				.map((child: ItemType<MenuItemType>) => filterNode(child))
				.filter((child: ItemType<MenuItemType> | null) => child !== null) as ItemType<MenuItemType>[];

			if (filteredChildren.length > 0 || selfMatch) {
				return {
					...node,
					...(filteredChildren.length > 0 && { children: filteredChildren }),
				} as ItemType<MenuItemType>;
			}
			return null;
		} else {
			return selfMatch ? node : null;
		}
	};
	return items.map(item => filterNode(item)).filter((item): item is MenuItemType => item !== null);
};

export const getAllMenuKeys = (items: TExtendedMenuItem[]): string[] => {
	const keys: string[] = [];

	const extractKeys = (menuItems: TExtendedMenuItem[]) => {
		menuItems.forEach(item => {
			if (item?.key) {
				keys.push(String(item.key));
			}
			// Type assertion for children property since MenuItemType might have children
			const itemWithChildren = item as MenuItemType & { children?: MenuItemType[] };
			if (itemWithChildren?.children && Array.isArray(itemWithChildren.children)) {
				extractKeys(itemWithChildren.children);
			}
		});
	};

	extractKeys(items);
	return keys;
};

export const dataFromLocalStorage = (key: ELocalStorageKeys) => {
	try {
		const item = localStorage.getItem(key);
		return item ? item : null;
	} catch (e) {
		console.error(`Error leyendo localStorage[${key}]:`, e);
		return null;
	}
};

export const getValidateStatus = (msg?: string) => (msg ? 'error' : undefined);