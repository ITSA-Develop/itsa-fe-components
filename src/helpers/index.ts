import {
	isArray,
	isEmpty,
	isNumber,
	isObject,
	isString,
	isUndefined,
	isValidEmailFormat,
	isValidMaskFormat,
} from './booleansChecks';
import { getInstance, securityApi } from './configAxios';
import { getFormatCurrency } from './currencies';
import { getCurrentDate } from './dates';
import {
	getDropdownFormattedOptions,
	getFormattedAddress,
	getFormattedFieldValue,
	getFormattedSchemaValues,
	getLabelFromValue,
	getSchemaColumns,
	getSchemaFields,
	getNumberFromStorage,
	getStringFromStorage,
} from './formats';
import {
	cleanObject,
	createValueLabelMap,
	filterMenuSidebar,
	filterMenuTree,
	findLastChild,
	generateFilteredMenu,
	getAllMenuKeys,
	getMappedOptionFromAPI,
	getMappedOptionFromLocal,
	getNestedValue,
	hasAllRequiredFields,
	mapPermissionsToMenuFormat,
} from './objects';
import { transformModuleToMenuData } from './menu/menuDataTransformer';
import { clearLocalStorage } from './session';
import { capitalize, isRouteActive, toTitleCase, getMicroFrontendByPath, joinUrl, isNumericOnly, filterNumericOnly, filterPositiveNumbersOnly } from './strings';
import {
	addOrderParamIfMissing,
	buildQueryParams,
	cleanPath,
	clearURLParams,
	fillRoute,
	getParamsFromURL,
	getQueryParam,
	updateURLParams,
} from './urls';

export {
	addOrderParamIfMissing,
	buildQueryParams,
	capitalize,
	cleanObject,
	cleanPath,
	clearLocalStorage,
	clearURLParams,
	createValueLabelMap,
	fillRoute,
	filterMenuSidebar,
	filterMenuTree,
	filterNumericOnly,
	filterPositiveNumbersOnly,
	findLastChild,
	getNumberFromStorage,
	getStringFromStorage,
	getInstance,
	generateFilteredMenu,
	getAllMenuKeys,
	getCurrentDate,
	getDropdownFormattedOptions,
	getFormatCurrency,
	getFormattedAddress,
	getFormattedFieldValue,
	getFormattedSchemaValues,
	getLabelFromValue,
	getMappedOptionFromAPI,
	getMappedOptionFromLocal,
	getNestedValue,
	getParamsFromURL,
	getQueryParam,
	getSchemaColumns,
	getSchemaFields,
	getMicroFrontendByPath,
	hasAllRequiredFields,
	isArray,
	isEmpty,
	isNumber,
	isNumericOnly,
	isObject,
	isRouteActive,
	isString,
	isUndefined,
	isValidEmailFormat,
	isValidMaskFormat,
	joinUrl,
	mapPermissionsToMenuFormat,
	securityApi,
	toTitleCase,
	transformModuleToMenuData,
	updateURLParams,
};
