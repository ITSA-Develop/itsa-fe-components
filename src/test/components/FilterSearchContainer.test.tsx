import { fireEvent, render, screen } from '@testing-library/react';
import { vi } from 'vitest';
import { FilterSearchContainer } from '../../components/FilterSearchContainer/FilterSearchContainer';

vi.mock('@/helpers', () => ({
	clearURLParams: vi.fn(),
}));

import { clearURLParams } from '@/helpers';

describe('FilterSearchContainer', () => {
	it('renders children and buttons', () => {
		const { container } = render(
			<FilterSearchContainer handleSubmit={vi.fn()} reset={vi.fn()}>
				<div>Child content</div>
			</FilterSearchContainer>,
		);

		expect(screen.getByText('Child content')).toBeInTheDocument();
		expect(screen.getByRole('button', { name: /limpiar filtros/i })).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it('calls handleSubmit on form submit', () => {
		const handleSubmit = vi.fn(e => e.preventDefault());
		const { container } = render(
			<FilterSearchContainer handleSubmit={handleSubmit} reset={vi.fn()}>
				<div />
			</FilterSearchContainer>,
		);

		const form = container.querySelector('form');
		expect(form).not.toBeNull();

		if (form) {
			fireEvent.submit(form);
		}

		expect(handleSubmit).toHaveBeenCalled();
	});

	it('calls reset and clearURLParams on Clear Filters button click', () => {
		const reset = vi.fn();
		render(
			<FilterSearchContainer handleSubmit={vi.fn()} reset={reset}>
				<div />
			</FilterSearchContainer>,
		);

		fireEvent.click(screen.getByRole('button', { name: /limpiar filtros/i }));
		expect(reset).toHaveBeenCalled();
		expect(clearURLParams).toHaveBeenCalled();
	});
});
