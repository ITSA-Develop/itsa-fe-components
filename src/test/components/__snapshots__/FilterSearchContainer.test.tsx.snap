// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FilterSearchContainer > renders children and buttons 1`] = `
<div>
  <form>
    <div
      class="p-4 w-full bg-grey-100 rounded-md"
    >
      <div
        class="flex flex-wrap gap-4 items-center justify-between"
      >
        <div>
          Child content
        </div>
        <button
          class="ant-btn css-dev-only-do-not-override-l9pxc0 ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
          type="button"
        >
          <span>
            Limpiar Filtros
          </span>
        </button>
        <button
          class="ant-btn css-dev-only-do-not-override-l9pxc0 ant-btn-default ant-btn-color-default ant-btn-variant-outlined !hidden"
          type="submit"
        />
      </div>
    </div>
  </form>
</div>
`;
